### PR TITLE
[wip] fix horovod installation

### LINKED ISSUE #12314: AttributeError: module 'horovod.torch' has no attribute 'nccl_built'
The `PL.pytorch-lightning (GPUs)` CI job is failing with `AttributeError: module 'horovod.torch' has no attribute 'nccl_built'`

https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=61024&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=059bb337-5de1-5e6a-4c56-42f55a328c0e

cc @tchaton @rohitgr7 @akihironitta @carmocca @borda

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Follows up #12368 and closes #12314.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: dockers/base-cuda/Dockerfile [hunk 1]
```diff
@@ -114,6 +114,10 @@ RUN \
    cat ./requirements/horovod.txt && \
    cmake --version && \
    pip install --no-cache-dir -r ./requirements/horovod.txt && \
+   horovodrun --check-build && \
+   pip uninstall -y horovod && \
+   pip install --no-cache-dir -r ./requirements/horovod.txt && \
+   horovodrun --check-build && \
```
Comment (akihironitta) [inline dockers/base-cuda/Dockerfile:120]: Reinstalling it somehow fixes the error...🤔

<details><summary>result here</summary>

```console
root@b43ef4616cac:/# horovodrun --check-build
Horovod v0.24.2:

Available Frameworks:
    [ ] TensorFlow
    [X] PyTorch
    [ ] MXNet

Available Controllers:
    [ ] MPI
    [ ] Gloo

Available Tensor Operations:
    [ ] NCCL
    [ ] DDL
    [ ] CCL
    [ ] MPI
    [ ] Gloo
root@b43ef4616cac:/# python -c "import horovod.torch"
Extension horovod.torch has not been built: /usr/local/lib/python3.7/dist-packages/horovod/torch/mpi_lib/_mpi_lib.cpython-37m-x86_64-linux-gnu.so not found
If this is not expected, reinstall Horovod with HOROVOD_WITH_PYTORCH=1 to debug the build error.
Warning! MPI libs are missing, but python applications are still available.
root@b43ef4616cac:/# pip uninstall -y horovod
...
root@b43ef4616cac:/# pip install --no-cache-dir horovod
...
root@b43ef4616cac:/# horovodrun --check-build
Horovod v0.24.2:

Available Frameworks:
    [ ] TensorFlow
    [X] PyTorch
    [ ] MXNet

Available Controllers:
    [ ] MPI
    [X] Gloo

Available Tensor Operations:
    [X] NCCL
    [ ] DDL
    [ ] CCL
    [ ] MPI
    [X] Gloo
root@b43ef4616cac:/# python -c "import horovod.torch"
root@b43ef4616cac:/#
```

</details>

Comment (awaelchli) [reply dockers/base-cuda/Dockerfile:120]: @akihironitta at this point, anything that makes it work is good enough :)) 
Did you push the image? You may need to retrigger the GPU CI job to pick up the changes

Comment (akihironitta) [reply dockers/base-cuda/Dockerfile:120]: Working on it!

**Comment by akihironitta** [comment]
TODO
- [ ] Confirm that the built docker image is pushed to the hub: https://github.com/PyTorchLightning/pytorch-lightning/runs/5590124349?check_suite_focus=true
- [ ] Revert FIXME in the nightly release CI config
- [ ] Confirm that the GPU testing passes in Azure CI
- [ ] Mark this PR as ready for review

File: .github/workflows/ci_dockers.yml [hunk 2]
```diff
@@ -43,7 +43,7 @@ jobs:
             PYTORCH_VERSION=${{ matrix.pytorch_version }}
           file: dockers/release/Dockerfile
           push: false
-        timeout-minutes: 50
+        timeout-minutes: 70  # FIXME
```
Comment (akihironitta) [inline .github/workflows/ci_dockers.yml:46]: Reviewers: Reinstalling horovod requires extra time, and 70 seems very reasonable as seen in https://github.com/PyTorchLightning/pytorch-lightning/runs/5590124760?check_suite_focus=true.

**Comment by akihironitta** [comment]
The built image has been successfully pushed to the hub, and GPU CI doesn't have the issue #12314 anymore, but I'm seeing another issue...

```
tests/callbacks/test_quantization.py::test_quantization[True-True-average] /__w/_temp/6c6a26ed-fe5c-4e9d-83d9-a3a6a6b6dabb.sh: line 1:   343 Segmentation fault      (core dumped) python -m coverage run --source pytorch_lightning -m pytest pytorch_lightning tests --ignore tests/benchmarks -v --junitxml=/__w/1/a/test-results.xml --durations=50
##[error]Bash exited with code '139'.
```

https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=61463&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=8b07f9df-ad34-5ead-2e0d-a2d35ff7ad3a

**Comment by akihironitta** [comment]
UPDATE: Couldn't reproduce the above error in our cluster. Will try to debug it in the CI...

**Comment by awaelchli** [comment]
@akihironitta maybe we try to skip the test to see if it is this particular test itself causing the problem or something else.

**Comment by Borda** [comment]
seems I found the bug, it was my typo in the past missing check for correct build https://github.com/PyTorchLightning/pytorch-lightning/pull/12318#discussion_r829785780

**Pull request closed without merging**